File: 01_Language/01_Functions/python/array_intersect.py
# coding: utf-8

import functools


def array_intersect(arr, arr2, *args):
    return functools.reduce(lambda x, y: x.intersection(y), [set(arr), arr2, *args])


if __name__ == '__main__':
    array1 = ["green", "red", "blue"]
    array2 = ["green", "yellow", "red"]
    print(array_intersect(array1, array2))

File: 01_Language/01_Functions/js/array_intersect.js
function array_intersect(arr) {
    let res = {}
    for (let i in arr) {
        let exists = false
        for (let j = 1; j < arguments.length; j++) {
            exists = false
            for (let k in arguments[j]) {
                if (arr[i] === arguments[j][k]) {
                    exists = true;
                    break
                }
            }
            if (!exists) {
                break
            }
        }
        if (exists) {
            res[i] = arr[i]
        }
    }
    return res
}

let array1 = ["green", "red", "blue"]
let array2 = ["green", "yellow", "red"]
console.log(array_intersect(array1, array2));
